service/worker/deployment/workflow.go

```diff
@@ -0,0 +1,174 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"time"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	sdkclient "go.temporal.io/sdk/client"
+	sdklog "go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/temporal"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	TaskQueue struct {
+		Name          string
+		TaskQueueType enumspb.TaskQueueType
+	}
+	Deployment struct {
+		Namespace      string
+		DeploymentName string
+		TaskQueue      []*TaskQueue // All the task queues associated with this buildID/deployment
+		BuildID        string
+	}
+	DeploymentName struct {
+		Name           string
+		CurrentBuildID string // denotes the current "default" build-ID of a deploymentName
+	}
+
+	// DeploymentWorkflowRunner holds the local state while running a deployment workflow
+	DeploymentWorkflowRunner struct {
+		deployment Deployment
+		ctx        workflow.Context
+		a          *activities
+		logger     sdklog.Logger
+		metrics    sdkclient.MetricsHandler
+	}
+)
+
+var (
+	defaultActivityOptions = workflow.ActivityOptions{
+		ScheduleToCloseTimeout: 1 * time.Hour,
+		StartToCloseTimeout:    30 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+			MaximumInterval: 60 * time.Second,
+		},
+	}
+)
+
+const continueAsNewThreshold = 1000
+
+func DeploymentWorkflow(ctx workflow.Context, deploymentWorkflowArgs Deployment) error {
+	deploymentWorkflowRunner := &DeploymentWorkflowRunner{
+		deployment: deploymentWorkflowArgs,
+		ctx:        ctx,
+		logger:     sdklog.With(workflow.GetLogger(ctx), "wf-namespace", deploymentWorkflowArgs.Namespace),
+		metrics:    workflow.GetMetricsHandler(ctx).WithTags(map[string]string{"namespace": deploymentWorkflowArgs.Namespace}),
+	}
+	// TQ should be set to nil here since signal handler updates it
+	deploymentWorkflowRunner.deployment.TaskQueue = nil
+	return deploymentWorkflowRunner.run()
+}
+
+func (d *DeploymentWorkflowRunner) run() error {
+
+	// Query Handler
+	err := workflow.SetQueryHandler(d.ctx, "list-task-queues", func(input []byte) ([]*TaskQueue, error) {
+		return d.deployment.TaskQueue, nil
+	})
+	if err != nil {
+		d.logger.Info("SetQueryHandler failed for list-task-queue with the error: " + err.Error())
+		return err
+	}
+
+	updateDeploymentSignalChannel := workflow.GetSignalChannel(d.ctx, "update_deployment")
+	updateBuildIDSignalChannel := workflow.GetSignalChannel(d.ctx, "update_task_queue_build_id")
```

Comment:
make these signal names package-level constants, and exported too. we'll need to reference them from frontend and elsewhere